crates/bevy_scene/src/dynamic_scene_builder.rs

```diff
@@ -0,0 +1,230 @@
+use crate::{DynamicEntity, DynamicScene};
+use bevy_app::AppTypeRegistry;
+use bevy_ecs::{prelude::Entity, reflect::ReflectComponent, world::World};
+use bevy_utils::{default, HashMap};
+
+/// A [`DynamicScene`] builder, used to build a scene from a [`World`] by extracting some entities.
+///
+/// ```
+/// # use bevy_scene::DynamicSceneBuilder;
+/// # use bevy_app::AppTypeRegistry;
+/// # use bevy_ecs::{
+/// #     component::Component, prelude::Entity, query::With, reflect::ReflectComponent, world::World,
+/// # };
+/// # use bevy_reflect::Reflect;
+/// # #[derive(Component, Reflect, Default, Eq, PartialEq, Debug)]
+/// # #[reflect(Component)]
+/// # struct ComponentA;
+/// # let mut world = World::default();
+/// # world.init_resource::<AppTypeRegistry>();
+/// # let entity = world.spawn(ComponentA).id();
+/// let mut builder = DynamicSceneBuilder::from_world(&world);
+/// builder.extract_entity(entity);
+/// let dynamic_scene = builder.build();
+/// ```
+pub struct DynamicSceneBuilder<'w> {
+    scene: HashMap<u32, DynamicEntity>,
+    type_registry: AppTypeRegistry,
+    world: &'w World,
+}
+
+impl<'w> DynamicSceneBuilder<'w> {
+    /// Prepare a builder that will extract entities and their component from the given [`World`].
+    /// All components registered in that world's [`AppTypeRegistry`] will be extracted.
+    pub fn from_world(world: &'w World) -> Self {
+        Self {
+            scene: default(),
+            type_registry: world.resource::<AppTypeRegistry>().clone(),
+            world,
+        }
+    }
+
+    /// Prepare a builder that will extract entities and their component from the given [`World`].
+    /// Only components registered in the given [`AppTypeRegistry`] will be extracted.
+    pub fn from_world_with_type_registry(world: &'w World, type_registry: AppTypeRegistry) -> Self {
+        Self {
+            scene: default(),
+            type_registry,
+            world,
+        }
+    }
+
+    /// Consume the builder, and produce the [`DynamicScene`].
+    pub fn build(self) -> DynamicScene {
+        DynamicScene {
+            entities: self.scene.into_values().collect(),
+        }
+    }
+
+    /// Extract one entity from the builder's [`World`].
+    pub fn extract_entity(&mut self, entity: Entity) -> &mut Self {
+        self.extract_entities(std::iter::once(entity))
+    }
+
+    /// Extract entities from the builder's [`World`].
+    ///
+    /// Extracting entities can be used to extract entities from a query:
```

Comment:
These docs should mention the behavior of extracting duplicate entities.

Reply:
Not completely convinced this should be mentioned, it would actually be very illogical to add the same entity twice in a scene. And pretty much a sure way to have a bug

Reply:
Hmm. I've often stored collections of entities in various resources. If I knew that this was idempotent WRT duplicate entities I wouldn't bother doing a uniqueness check before passing that collection into this method.

Reply:
I think I agree with @alice-i-cecile that it should let the user know that they don't need to filter for duplicate entities before using this method. I can see why a user might think duplicates are possible, seeing as the [scene format](https://github.com/bevyengine/bevy/blob/9a597b758e5b2b2cc4355185a3fcba9564bfb104/assets/scenes/load_scene_example.scn.ron) stores entities in an array rather than a map.

Reply:
👍